src/pykeen/quadruples/quadruples_factory.py

```diff
@@ -0,0 +1,1246 @@
+        timestamp_column = timestamp_column[~non_mappable_quadruples, None]
+        logger.warning(
+            f"In total {non_mappable_quadruples.sum():.0f} from {quadruples.shape[0]:.0f} quadruples were filtered out",
+        )
+    quadruples_of_ids = np.concatenate(
+        [head_column, relation_column, tail_column, timestamp_column], axis=1
+    )
+    quadruples_of_ids = np.array(quadruples_of_ids, dtype=np.int64)
+    unique_mapped_quadruples = np.unique(ar=quadruples_of_ids, axis=0)
+    return torch.tensor(unique_mapped_quadruples, dtype=torch.long)
+
+
+class CoreQuadruplesFactory(KGInfo):
+    """Create instances from ID-based quadruples."""
+
+    def __init__(
+        self,
+        mapped_quadruples: Union[MappedQuadruples, np.ndarray],
+        num_entities: int,
+        num_relations: int,
+        num_timestamps: int,
+        create_inverse_quadruples: bool = False,
+        metadata: Optional[Mapping[str, Any]] = None,
+    ):
+        """
+        Create the quadruples factory.
+
+        :param mapped_quadruples: shape: (n, 4)
+            A four-column matrix where each row are the head identifier, relation identifier, tail identifier, then timestamp.
+        :param num_entities:
+            The number of entities.
+        :param num_relations:
+            The number of relations.
+        :param num_timestamps:
+            The number of timestamps.
+        :param create_inverse_quadruples:
+            Whether to create inverse quadruples.
+        :param metadata:
+            Arbitrary metadata to go with the graph.
+
+        :raises TypeError:
+            if the mapped_quadruples are of non-integer dtype
+        :raises ValueError:
+            if the mapped_quadruples are of invalid shape
+        """
+        super().__init__(
+            num_entities=num_entities,
+            num_relations=num_relations,
+            create_inverse_triples=create_inverse_quadruples,
+        )
+        self.create_inverse_quadruples = create_inverse_quadruples
+        self.num_timestamps = num_timestamps
+        # ensure torch.Tensor
+        mapped_quadruples = torch.as_tensor(mapped_quadruples)
+        # input validation
+        if mapped_quadruples.ndim != 2 or mapped_quadruples.shape[1] != 4:
+            raise ValueError(
+                f"Invalid shape for mapped_quadruples: {mapped_quadruples.shape}; must be (n, 3)"
+            )
+        if mapped_quadruples.is_complex() or mapped_quadruples.is_floating_point():
+            raise TypeError(f"Invalid type: {mapped_quadruples.dtype}. Must be integer dtype.")
+        # always store as torch.long, i.e., torch's default integer dtype
+        self.mapped_quadruples = mapped_quadruples.to(dtype=torch.long)
+        if metadata is None:
+            metadata = dict()
+        self.metadata = metadata
+        self.relation_inverter = relation_inverter_resolver.make(query=None)
+
+    @classmethod
```

Comment:
(I'll skip annotating the rest of the file - the overall pattern seems to be that there is quite a bit of code duplication with `CoreTriplesFactory` and it is worth to explore how to reduce that)